gradio/processing_utils.py

```diff
@@ -271,36 +273,71 @@ def save_file_to_cache(file_path: str | Path, cache_dir: str) -> str:
     return full_temp_file_path
 
 
-def check_public_url(url: str):
+@lru_cache(maxsize=256)
+def resolve_with_google_dns(hostname: str) -> str | None:
+    url = f"https://dns.google/resolve?name={hostname}&type=A"
+
+    with urllib.request.urlopen(url) as response:
+        data = json.loads(response.read().decode())
+
+    if data.get("Status") == 0 and "Answer" in data:
+        for answer in data["Answer"]:
+            if answer["type"] == 1:
+                return answer["data"]
+
+
+# Always return these URLs as is, without checking to see if they resolve
+# to an internal IP address. This is because Hugging Face uses DNS splitting,
+# which means that requests from HF Spaces to HF Datasets, Models, or Spaces
+# may resolve to internal IP addresses even if they are publicly accessible.
+PUBLIC_URL_WHITELIST = ["hf.co", "huggingface.co", "hf.space"]
```

Comment:
Is `hf.space` actually expected? The hostname of a space is `<user>-<space-id>.hf.space` which contains `hf.space` but would not be in the whitelist.

Reply:
Good point! I'll remove that one, I actually don't even think we need to whitespace Spaces, its just models or datasets